### PR TITLE
Software creation

### PR DESCRIPTION
**Pull request by shthiago**

### CODE REVIEW

File: squad_3_ad_data_science/feature_engineering.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                                drop_first=True)
+
+    # Scale data
+    scaler = StandardScaler()
+    input_data[:] = scaler.fit_transform(input_data)
+
+    return input_data
+
+
+def impute_as_config(input_data: pd.DataFrame):
```
Comment (mfrata) [inline squad_3_ad_data_science/feature_engineering.py:149]: typo

File: squad_3_ad_data_science/feature_engineering.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                raise KeyError
+
+        except KeyError:
+            raise ConfigMissException(f'config.NAN_FIXES is with some value ' +
+                                      f'error on col {col} with info = {info}')
+
+    return input_data
+
+
+class ConfigMissException(Exception):
```
Comment (mfrata) [inline squad_3_ad_data_science/feature_engineering.py:194]: this should be on a separated file called `exceptions.py` or like in [here](https://github.com/pandas-dev/pandas/blob/master/pandas/errors/__init__.py)

File: squad_3_ad_data_science/main.py [hunk 3]
```diff
@@ -25,6 +40,48 @@ def features(**kwargs):
...
+
+    logger.info('Loading data')
+    pure_data = pd.read_csv(config.train_dataset, index_col=0)
+
+    old_shape = pure_data.shape
+    logger.info('Working on features')
+    try:
+        data = pipeline_data(pure_data)
+    except Exception:
+        logger.error('Something where wrong')
```
Comment (mfrata) [inline squad_3_ad_data_science/main.py:69]: "something went wrong"

File: squad_3_ad_data_science/main.py [hunk 4]
```diff
@@ -25,6 +40,48 @@ def features(**kwargs):
...
+        exit(1)
+
+    logger.info('Selecting variables with PCA')
+
+    data = select_features(data)
+
+    logger.info(f'Data original shape: {old_shape}')
+    logger.info(f'Data treining shape: {data.shape}')
```
Comment (mfrata) [inline squad_3_ad_data_science/main.py]: typo

File: squad_3_ad_data_science/main.py [hunk 5]
```diff
@@ -67,6 +146,78 @@ def metadata(**kwargs):
...
+    # Run over each test file, loading its ids and running the prediction
+    for f in config.test_datasets:
+        if not isfile(f):
+            logger.error(f'Test file `{f}` not find. Ignoring...`')
+            continue
+
+        logger.info(f'Processing file `{f}`...')
+
+        # Label to identify register
+        name = f.split('/')[-1][:-4]
```
Comment (mfrata) [inline squad_3_ad_data_science/main.py:188]: What are you trying to get here?

File: squad_3_ad_data_science/main.py [hunk 6]
```diff
@@ -67,6 +146,78 @@ def metadata(**kwargs):
...
+
+    else:
+        update = True
+
+    # Load saved labeled data
+    labeld_data_pkl_path = join(config.data_path,
+                                config.TRAIN_DATA_PKL_LABELED)
+    data = pd.read_pickle(labeld_data_pkl_path)
+
+    # TODO apply Nathan's function to calculate Average PRecision for each csv on tests
```
Comment (mfrata) [inline squad_3_ad_data_science/main.py]: don't forget this...

File: Pipfile [hunk 7]
```diff
@@ -11,9 +10,15 @@ numpy = "*"
 matplotlib = "*"
 pandas = "*"
 jupyter = "*"
-jinja = "*"
 sklearn = "*"
 jinja2 = "*"
+flake8 = "*"
```
Comment (mfrata) [inline Pipfile:15]: always choose a specific version. Otherwise you can (in the future) break your software

File: squad_3_ad_data_science/config.py [hunk 8]
```diff
@@ -22,7 +25,7 @@
 ]
 
 # test_size for metadate generation
```
Comment (mfrata) [inline squad_3_ad_data_science/config.py:27]: typo

File: squad_3_ad_data_science/main.py [hunk 9]
```diff
@@ -65,17 +64,24 @@ def features(**kwargs):
...
+    logger.info(f'Selected {len(data.columns)-1} features for' +
+                f' a {config.EXPLAINED_VARIANCE_RATIO*100}% variance' +
+                f' representation')
+    logger.info(f'Concatenating special labels')
+
+    data = data.join(special_labels_df)
 
     logger.info(f'Data original shape: {old_shape}')
-    logger.info(f'Data treining shape: {data.shape}')
+    logger.info(f'Data training shape: {data.shape}')
```
Comment (mfrata) [inline squad_3_ad_data_science/main.py:84]: ha you fixed!

File: squad_3_ad_data_science/main.py [hunk 10]
```diff
@@ -219,15 +226,76 @@ def metadata(**kwargs):
                 f'available in `{config.performance_metadata_path}`')
 
 
-def predict(input_data):
+def predict(input_data, **kwargs):
     """Predict: load the trained model and score input_data
 
+        @kwargs:
```
Comment (mfrata) [inline squad_3_ad_data_science/main.py:232]: great doc

File: tests/test_feature_engineering.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
+import pandas as pd
+
+from squad_3_ad_data_science.feature_engineering import pipeline_data, select_features
+from squad_3_ad_data_science import config
+
+def test_data(fixture_test_data):
+    assert type(fixture_test_data) == pd.DataFrame
+
+def test_pipeline_data(fixture_test_data):
```
Comment (mfrata) [inline tests/test_feature_engineering.py:9]: great tests!

File: squad_3_ad_data_science/feature_engineering.py (hunk 1 above)
Comment (shthiago) [reply squad_3_ad_data_science/feature_engineering.py:149]: where? '-'

File: squad_3_ad_data_science/feature_engineering.py (hunk 2 above)
Comment (shthiago) [reply squad_3_ad_data_science/feature_engineering.py:194]: And what is the best option? on `__init__. py` looks graceful, but on another file like `exceptions.py` seems to be more clear...

File: squad_3_ad_data_science/main.py (hunk 5 above)
Comment (shthiago) [reply squad_3_ad_data_science/main.py:188]: Looping over three test files (`estaticos_market<1,2,3>.csv`) to generate performance metadata with our validation methods (currently just Average Precision)

Bad idea?

File: Pipfile (hunk 7 above)
Comment (shthiago) [reply Pipfile:15]: Oh, forgot about it. I will fix it soon.

File: squad_3_ad_data_science/config.py (hunk 8 above)
Comment (shthiago) [reply squad_3_ad_data_science/config.py:27]: I will fix it too.

File: squad_3_ad_data_science/main.py (hunk 3 above)
Comment (shthiago) [reply squad_3_ad_data_science/main.py:69]: Oops. Forget about it. Previous errors log should clarify what happened. I will rewrite this log too.

File: squad_3_ad_data_science/main.py (hunk 5 above)
Comment (shthiago) [reply squad_3_ad_data_science/main.py:188]: Or are you talking about line 184? There I'm trying to get a label to identify the performance record, removing `.csv` from the end of filename

File: squad_3_ad_data_science/feature_engineering.py (hunk 1 above)
Comment (mfrata) [reply squad_3_ad_data_science/feature_engineering.py:149]: I thought u meant `input`

File: squad_3_ad_data_science/feature_engineering.py (hunk 2 above)
Comment (mfrata) [reply squad_3_ad_data_science/feature_engineering.py:194]: There isn't a *best* option. But on `exceptions.py` is clear (at least to me) where all the custom exceptions are declared.

File: squad_3_ad_data_science/main.py (hunk 5 above)
Comment (mfrata) [reply squad_3_ad_data_science/main.py:188]: os.path lib have these functions ready. Now that you commented your intention ( you could have commented it right above the line). For instance:
````pyhton
>>> import os
>>> base=os.path.basename('/root/dir/sub/file.ext')
>>> base
'file.ext'
>>> os.path.splitext(base)
('file', '.ext')
>>> os.path.splitext(base)[0]
'file'
````

source: https://stackoverflow.com/a/678266/7299191
